File: sdk/python/kms_paramstore/values.py
"""Declarative, store-backed config fields (the Python descriptor idiom).

Declare :class:`SecretValue` / :class:`ParameterValue` as class attributes on a
config object, then resolve them all with a single :meth:`Client.resolve` call::

    class AppConfig:
        stripe_key = SecretValue("stripe-api-key", token="...")
        rate_limit = ParameterValue("rate-limit")              # hot-reloads
        log_format = ParameterValue("log-format", static=True)  # boot-time only

    cfg = AppConfig()
    client.resolve(cfg)

    cfg.stripe_key            # -> a redacting Secret
    cfg.stripe_key.value      # -> bytes plaintext (explicit access only)
    print(cfg.stripe_key)     # -> [REDACTED]
    cfg.rate_limit.get()      # -> latest value, hot-reloaded (default)

Keys are relative to the client namespace, or absolute ``/env/app/key``.
Parameters hot-reload by default; pass ``static=True`` for a boot-time-only read.

Resolution order for each field: environment override, then store fetch, then
``default``, else an error naming the key (fail fast at startup).
"""

from __future__ import annotations

import functools
import os
import threading
import weakref
from concurrent.futures import ThreadPoolExecutor
from typing import TYPE_CHECKING, Callable, Dict, List, Optional, Tuple

from . import errors
from .secret import Secret

if TYPE_CHECKING:
    from .client import Client

__all__ = ["SecretValue", "ParameterValue", "resolve_targets"]


# ---------------------------------------------------------------------------
# Per-instance mutable state
# ---------------------------------------------------------------------------


class _SecretState:
    __slots__ = ("lock", "initialized", "from_env", "secret")

    def __init__(self) -> None:
        self.lock = threading.RLock()
        self.initialized = False
        self.from_env = False
        self.secret = Secret(b"")


class _ParamState:
    __slots__ = ("lock", "initialized", "from_env", "static", "default", "client", "value", "callbacks")

    def __init__(self) -> None:
        self.lock = threading.RLock()
        self.initialized = False
        self.from_env = False
        self.static = False
        self.default: Optional[str] = None
        self.client: Optional["Client"] = None
        self.value = ""
        self.callbacks: List[Callable[[str, str], None]] = []

    def apply_update(self, new_val: str, present: bool) -> None:
        # A deletion reverts to the configured default so the app stops serving a
        # value the store no longer has. With no default the last-known value is
        # kept, since apps rarely want config to vanish underneath them.
        if not present:
            if self.default is None:
                return
            new_val = self.default
        with self.lock:
            old = self.value
            if old == new_val:
                return
            self.value = new_val
            cbs = list(self.callbacks)
            client = self.client
        if client is None:
            return
        for cb in cbs:
            client._enqueue_callback(functools.partial(cb, old, new_val))


# ---------------------------------------------------------------------------
# Parameter handle returned by descriptor access
# ---------------------------------------------------------------------------


class ParameterHandle:
    """Handle for a resolved :class:`ParameterValue`; hot-reloads unless static."""

    __slots__ = ("_key", "_state")

    def __init__(self, key: str, state: _ParamState) -> None:
        self._key = key
        self._state = state

    @property
    def initialized(self) -> bool:
        with self._state.lock:
            return self._state.initialized

    def get(self) -> str:
        """Return the latest value (``""`` before resolution)."""
        return self._state.value

    @property
    def value(self) -> str:
        return self._state.value

    def on_change(self, fn: Callable[[str, str], None]) -> None:
        """Register a callback ``fn(old, new)`` fired on a dispatch thread.

        A callback on a static or env-pinned value never fires.
        """
        with self._state.lock:
            self._state.callbacks.append(fn)

    def __repr__(self) -> str:
        return f"ParameterHandle({self._key!r}={self._state.value!r})"

    def __str__(self) -> str:
        return self._state.value


# ---------------------------------------------------------------------------
# Descriptors
# ---------------------------------------------------------------------------


class _DescriptorBase:
    def __init__(self) -> None:
        self._name = ""
        self._state_key = ""
        # Fallback store for instances without a usable __dict__.
        self._weak: "weakref.WeakKeyDictionary[object, object]" = weakref.WeakKeyDictionary()

    def __set_name__(self, owner: type, name: str) -> None:
        self._name = name
        self._state_key = f"_kms_paramstore_state_{name}"

    def _new_state(self) -> object:
        raise NotImplementedError

    def _state_for(self, instance: object) -> object:
        # Prefer the instance __dict__ (no weakref requirement); fall back to a
        # WeakKeyDictionary for slotted instances.
        d = getattr(instance, "__dict__", None)
        if d is not None:
            st = d.get(self._state_key)
            if st is None:
                st = self._new_state()
                d[self._state_key] = st
            return st
        st = self._weak.get(instance)
        if st is None:
            st = self._new_state()
            self._weak[instance] = st
        return st

    def _init(self, instance: object, client: "Client", *, timeout: Optional[float] = None) -> None:
        raise NotImplementedError


class SecretValue(_DescriptorBase):
    """A declarative, store-backed secret field that resolves to a :class:`Secret`.

    Args:
        key: relative key (``"stripe-api-key"``) or absolute ``"/env/app/key"``.
        token: per-secret access token; for client-bound secrets it is also the
            client key share.
        env_var: optional environment variable that, when set and non-empty,
            overrides the store value (no namespace resolution is needed then).
        default: optional fallback (development only).
    """

    def __init__(self, key: str = "", *, token: Optional[str] = None, env_var: Optional[str] = None,
                 default: Optional[str] = None) -> None:
        super().__init__()
        self._key = key
        self._token = token or ""
        self._env_var = env_var or ""
        self._default = default

    def _new_state(self) -> object:
        return _SecretState()

    def __get__(self, instance: object, owner: Optional[type] = None) -> "SecretValue | Secret":
        if instance is None:
            return self
        st = self._state_for(instance)
        assert isinstance(st, _SecretState)
        with st.lock:
            if not st.initialized:
                raise errors.NotInitializedError(f"secret {self._key!r} read before resolve")
            return st.secret

    def _init(self, instance: object, client: "Client", *, timeout: Optional[float] = None) -> None:
        st = self._state_for(instance)
        assert isinstance(st, _SecretState)
        with st.lock:
            if st.initialized:
                return
            if self._env_var:
                ev = os.environ.get(self._env_var, "")
                if ev != "":
                    st.secret = Secret(ev.encode("utf-8"), key=self._key)
                    st.from_env = True
                    st.initialized = True
                    client._logf("secret %r resolved from env %s (store fetch skipped)", self._key, self._env_var)
                    return
            if self._key:
                try:
                    sec = client.get_secret(self._key, secret_token=self._token, timeout=timeout)
                except Exception as err:
                    if self._default is not None and client._default_allowed_for_error(err):
                        client._logf("secret %r fetch failed (%s); using default", self._key, err)
                        st.secret = Secret(self._default.encode("utf-8"), key=self._key)
                        st.initialized = True
                        return
                    raise errors.ParamStoreError(f"resolve secret {self._key!r}: {err}") from err
                st.secret = sec
                st.initialized = True
                return
            if self._default is not None:
                st.secret = Secret(self._default.encode("utf-8"), key=self._key)
                st.initialized = True
                return
            raise errors.ConfigError("SecretValue has no key, env_var, or default configured")


class ParameterValue(_DescriptorBase):
    """A declarative, store-backed non-secret field.

    Hot reload is on by default: the value tracks the Subscribe stream and
    :meth:`ParameterHandle.get` always returns the latest value without an RPC.
    Every non-static field in a namespace shares that namespace's single
    subscription. Pass ``static=True`` for a boot-time-only read (no
    subscription). An
    env-pinned field never hot-reloads regardless of ``static``.

    Args:
        key: relative key (``"rate-limit"``) or absolute ``"/env/app/key"``.
        env_var: environment variable that, when set and non-empty, pins the
            value (no store fetch, no hot reload, no namespace resolution).
        default: optional fallback (development only).
        static: opt out of hot reload — read once at resolve time.
    """

    def __init__(self, key: str = "", *, env_var: Optional[str] = None, default: Optional[str] = None,
                 static: bool = False) -> None:
        super().__init__()
        self._key = key
        self._env_var = env_var or ""
        self._default = default
        self._static = static

    def _new_state(self) -> object:
        return _ParamState()

    def __get__(self, instance: object, owner: Optional[type] = None) -> "ParameterValue | ParameterHandle":
        if instance is None:
            return self
        st = self._state_for(instance)
        assert isinstance(st, _ParamState)
        return ParameterHandle(self._key, st)

    def _init(self, instance: object, client: "Client", *, timeout: Optional[float] = None) -> None:
        st = self._state_for(instance)
        assert isinstance(st, _ParamState)
        with st.lock:
            if st.initialized:
                return
            st.client = client
            st.default = self._default  # revert target for a later deletion
            if self._env_var:
                ev = os.environ.get(self._env_var, "")
                if ev != "":
                    st.value = ev
                    st.from_env = True
                    st.static = True
                    st.initialized = True
                    if not self._static:
                        client._logf("parameter %r pinned to env %s; hot reload disabled", self._key, self._env_var)
                    return
            ref, value = self._resolve_from_store(client, timeout)
            st.value = value
            st.initialized = True
            # Hot reload on by default: register on the namespace subscription.
            if not self._static and ref is not None:
                st.static = False
                client._subs().register_param(ref, value, st.apply_update)
            else:
                st.static = True

    def _resolve_from_store(self, client: "Client", timeout: Optional[float]):
        """Return ``(ref, value)``; ``ref`` is None only for a default-only descriptor."""
        if self._key:
            # A relative key on a namespace-less client is a config error naming
            # the key; it must propagate rather than fall back to a default.
            ref = client._resolve_ref(self._key)
            try:
                value = client._get_parameter_ref(ref, timeout=timeout)
            except Exception as err:
                if self._default is not None and client._default_allowed_for_error(err):
                    client._logf("parameter %r fetch failed (%s); using default", self._key, err)
                    # Keep the resolved ref so non-static values subscribe and
                    # can replace the fallback when the missing parameter is
                    # created later.
                    return ref, self._default
                raise errors.ParamStoreError(f"resolve parameter {self._key!r}: {err}") from err
            return ref, value
        if self._default is not None:
            return None, self._default
        raise errors.ConfigError("ParameterValue has no key, env_var, or default configured")


# ---------------------------------------------------------------------------
# Resolution walk
# ---------------------------------------------------------------------------


def _has_descriptor(cls: type) -> bool:
    for klass in cls.__mro__:
        for attr in vars(klass).values():
            if isinstance(attr, _DescriptorBase):
                return True
    return False


def _collect_targets(obj: object, targets: List[Tuple[_DescriptorBase, object]], visited: set) -> None:
    if obj is None or id(obj) in visited:
        return
    visited.add(id(obj))
    for klass in type(obj).__mro__:
        for attr in vars(klass).values():
            if isinstance(attr, _DescriptorBase):
                targets.append((attr, obj))
    inst_dict: Optional[Dict[str, object]] = getattr(obj, "__dict__", None)
    if inst_dict:
        for v in list(inst_dict.values()):
            if _is_config_like(v):
                _collect_targets(v, targets, visited)


def _is_config_like(v: object) -> bool:
    if v is None or isinstance(v, (str, bytes, bytearray, int, float, bool, complex, Secret, ParameterHandle)):
        return False
    return _has_descriptor(type(v))


def resolve_targets(client: "Client", config_obj: object, *, timeout: Optional[float] = None) -> None:
    """Resolve every declarative field on ``config_obj`` concurrently."""
    targets: List[Tuple[_DescriptorBase, object]] = []
    _collect_targets(config_obj, targets, set())
    if not targets:
        return

    errors_out: List[Optional[Exception]] = [None] * len(targets)

    def run(i: int, desc: _DescriptorBase, instance: object) -> None:
        try:
            desc._init(instance, client, timeout=timeout)
        except Exception as e:  # recorded and re-raised below
            errors_out[i] = e

    # The service exposes no batch-read RPC, so "batch into as few RPCs as
    # possible" means issuing the independent fetches concurrently.
    with ThreadPoolExecutor(max_workers=min(16, len(targets))) as pool:
        futures = [pool.submit(run, i, desc, inst) for i, (desc, inst) in enumerate(targets)]
        for f in futures:
            f.result()

    for e in errors_out:
        if e is not None:
            raise e

File: sdk/go/kmsclient/values.go
package kmsclient

import (
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"sync"
	"sync/atomic"
)

// SecretValue is a declarative, store-backed secret field. Declare it in a
// config struct, then resolve it with Init (or via Client.Resolve). Once
// initialized it redacts itself in all string/JSON formatting; plaintext is
// only reachable through Value/StringValue.
//
// The mutable state lives behind an unexported pointer so that SecretValue's
// redacting formatter methods can use value receivers (a struct that embedded a
// mutex directly could not, and would leak plaintext through fmt's reflection
// of unexported fields).
type SecretValue struct {
	// Key is the secret key, relative to the client namespace (e.g.
	// "stripe/api-key"), or an absolute "/env/app/key" display path.
	Key string
	// Token is the per-secret access token. For client-bound secrets it is also
	// the client key share (plan 10.7).
	Token string
	// EnvVar is an optional environment variable that, when set and non-empty,
	// overrides the store value.
	EnvVar string
	// Default is an optional fallback value, intended for development only.
	Default string

	state *secretState
}

type secretState struct {
	mu          sync.RWMutex
	initialized bool
	fromEnv     bool
	value       string
}

func (v *SecretValue) ensureState() *secretState {
	if v.state == nil {
		v.state = &secretState{}
	}
	return v.state
}

// Init resolves the value using the standard order: env override, then store
// fetch, then Default (only when the store reports the value absent, unless
// Config.FallbackToDefaultsOnError is set), else an error naming the path. It is
// idempotent: a second call after success is a no-op. Init is not safe to call
// concurrently on the same SecretValue (Client.Resolve initializes distinct
// fields in parallel, which is safe).
func (v *SecretValue) Init(client *Client) error {
	return v.InitContext(context.Background(), client)
}

// InitContext is Init with a caller-supplied context.
func (v *SecretValue) InitContext(ctx context.Context, client *Client) error {
	st := v.ensureState()
	st.mu.Lock()
	defer st.mu.Unlock()
	if st.initialized {
		return nil
	}

	if v.EnvVar != "" {
		if ev := os.Getenv(v.EnvVar); ev != "" {
			st.value = ev
			st.fromEnv = true
			st.initialized = true
			client.logf("kmsclient: secret %q resolved from env %s (store fetch skipped)", v.Key, v.EnvVar)
			return nil
		}
	}

	if v.Key != "" {
		sec, err := client.GetSecret(ctx, v.Key, WithSecretToken(v.Token))
		if err == nil {
			st.value = sec.StringValue()
			st.initialized = true
			return nil
		}
		// Default is a dev-only escape hatch: use it when the store reports the
		// value absent (ErrNotFound), but never mask a connectivity/auth error
		// with it unless the caller opted into FallbackToDefaultsOnError.
		if v.Default != "" && client.defaultAllowedForErr(err) {
			client.logf("kmsclient: secret %q fetch failed (%v); using Default", v.Key, err)
			st.value = v.Default
			st.initialized = true
			return nil
		}
		return fmt.Errorf("kmsclient: resolve secret %q: %w", v.Key, err)
	}

	if v.Default != "" {
		st.value = v.Default
		st.initialized = true
		return nil
	}
	return fmt.Errorf("kmsclient: secret has no Key, EnvVar, or Default configured")
}

// Value returns the resolved plaintext. It panics if the value has not been
// initialized, which surfaces wiring mistakes at first use rather than serving
// an empty secret.
func (v *SecretValue) Value() string {
	st := v.state
	if st == nil {
		panic(fmt.Sprintf("kmsclient: SecretValue(%q).Value() called before Init", v.Key))
	}
	st.mu.RLock()
	defer st.mu.RUnlock()
	if !st.initialized {
		panic(fmt.Sprintf("kmsclient: SecretValue(%q).Value() called before Init", v.Key))
	}
	return st.value
}

// StringValue is an alias for Value.
func (v *SecretValue) StringValue() string { return v.Value() }

// Initialized reports whether the value has been resolved.
func (v *SecretValue) Initialized() bool {
	st := v.state
	if st == nil {
		return false
	}
	st.mu.RLock()
	defer st.mu.RUnlock()
	return st.initialized
}

// Secret returns the resolved plaintext wrapped in a redacting Secret.
func (v *SecretValue) Secret() Secret { return NewSecret([]byte(v.Value())) }

// String redacts. Value receiver so both SecretValue and *SecretValue redact,
// and so fmt does not fall through to the unexported state.
func (v SecretValue) String() string { return redactedText }

// GoString redacts (used by %#v).
func (v SecretValue) GoString() string { return redactedText }

// Format redacts for every verb.
func (v SecretValue) Format(f fmt.State, verb rune) {
	switch verb {
	case 'q':
		_, _ = fmt.Fprintf(f, "%q", redactedText)
	default:
		_, _ = io.WriteString(f, redactedText)
	}
}

// MarshalJSON redacts.
func (v SecretValue) MarshalJSON() ([]byte, error) {
	return []byte(`"` + redactedText + `"`), nil
}

// ParameterValue is a declarative, store-backed non-secret field. By default it
// hot-reloads over the client's shared Subscribe stream, so Get always returns
// the latest value without an RPC. Set Static to opt out and resolve once at
// Init.
//
// Like SecretValue, mutable state lives behind a pointer so the value type
// stays copy-safe.
type ParameterValue struct {
	// Key is the parameter key, relative to the client namespace (e.g.
	// "rate-limit"), or an absolute "/env/app/key" display path.
	Key string
	// EnvVar optionally overrides the store value; an env-set value is pinned
	// and does not hot-reload.
	EnvVar string
	// Default is an optional fallback, intended for development.
	Default string
	// Static disables hot reload: the value resolves once at Init and never
	// changes afterward. The zero value keeps hot reload ON — every non-static
	// ParameterValue tracks its namespace over the shared Subscribe stream.
	Static bool

	state *paramState
}

type paramState struct {
	value       atomic.Pointer[string]
	mu          sync.Mutex
	initialized bool
	fromEnv     bool
	client      *Client
	callbacks   []func(old, new string)
}

func (p *ParameterValue) ensureState() *paramState {
	if p.state == nil {
		p.state = &paramState{}
	}
	return p.state
}

func storeString(ptr *atomic.Pointer[string], s string) { ptr.Store(&s) }

// Init resolves the value (env override, store fetch, Default only when the
// store reports the value absent unless Config.FallbackToDefaultsOnError is set,
// else error) and, unless Static is set or the value came from an env override,
// registers the value's namespace on the client's shared subscription for hot
// reload. It is idempotent.
func (p *ParameterValue) Init(client *Client) error {
	return p.InitContext(context.Background(), client)
}

// InitContext is Init with a caller-supplied context.
func (p *ParameterValue) InitContext(ctx context.Context, client *Client) error {
	st := p.ensureState()
	st.mu.Lock()
	defer st.mu.Unlock()
	if st.initialized {
		return nil
	}
	st.client = client

	if p.EnvVar != "" {
		if ev := os.Getenv(p.EnvVar); ev != "" {
			storeString(&st.value, ev)
			st.fromEnv = true
			st.initialized = true
			if !p.Static {
				client.logf("kmsclient: parameter %q pinned to env %s; hot reload disabled", p.Key, p.EnvVar)
			}
			return nil
		}
	}

	var (
		r       ref
		haveRef bool
	)
	if p.Key != "" {
		rr, err := client.resolveRef(ctx, p.Key)
		if err != nil {
			return err
		}
		r, haveRef = rr, true
	}

	value, err := p.resolveFromStore(ctx, client, r, haveRef)
	if err != nil {
		return err
	}
	storeString(&st.value, value)
	st.initialized = true

	if !p.Static && haveRef {
		client.subs().registerParam(r, value, p.applyUpdate)
	}
	return nil
}

func (p *ParameterValue) resolveFromStore(ctx context.Context, client *Client, r ref, haveRef bool) (string, error) {
	if haveRef {
		v, err := client.getParameter(ctx, r, getOptions{})
		if err == nil {
			return v, nil
		}
		// See SecretValue.InitContext: Default covers an absent value, not an
		// unreachable store, unless FallbackToDefaultsOnError is set.
		if p.Default != "" && client.defaultAllowedForErr(err) {
			client.logf("kmsclient: parameter %q fetch failed (%v); using Default", p.Key, err)
			return p.Default, nil
		}
		return "", fmt.Errorf("kmsclient: resolve parameter %q: %w", p.Key, err)
	}
	if p.Default != "" {
		return p.Default, nil
	}
	return "", errors.New("kmsclient: parameter has no Key, EnvVar, or Default configured")
}

// applyUpdate is invoked by the subscription manager when a new value arrives
// for this parameter's path. present is false for deletions: the value reverts
// to the configured Default so the app stops serving a value the store no longer
// has (defect M3). With no Default the last-known value is kept, since apps
// rarely want config to vanish underneath them.
func (p *ParameterValue) applyUpdate(newVal string, present bool) {
	st := p.state
	if st == nil {
		return
	}
	if !present {
		if p.Default == "" {
			return
		}
		newVal = p.Default
	}
	old := st.value.Swap(&newVal)
	oldVal := ""
	if old != nil {
		oldVal = *old
	}
	if oldVal == newVal {
		return
	}
	p.dispatch(oldVal, newVal)
}

func (p *ParameterValue) dispatch(oldVal, newVal string) {
	st := p.state
	st.mu.Lock()
	cbs := make([]func(old, new string), len(st.callbacks))
	copy(cbs, st.callbacks)
	client := st.client
	st.mu.Unlock()
	if client == nil {
		return
	}
	for _, cb := range cbs {
		cb := cb
		client.enqueueCallback(p.Key, func() { cb(oldVal, newVal) })
	}
}

// Get returns the latest value. Safe for concurrent use; returns "" before Init.
func (p *ParameterValue) Get() string {
	st := p.state
	if st == nil {
		return ""
	}
	if v := st.value.Load(); v != nil {
		return *v
	}
	return ""
}

// OnChange registers a callback invoked (on a dedicated goroutine) whenever the
// value changes. old and new are the previous and current values. Registering a
// callback on a Static or env-pinned value is allowed but it will never fire.
func (p *ParameterValue) OnChange(fn func(old, new string)) {
	st := p.ensureState()
	st.mu.Lock()
	st.callbacks = append(st.callbacks, fn)
	st.mu.Unlock()
}

// Initialized reports whether the value has been resolved.
func (p *ParameterValue) Initialized() bool {
	st := p.state
	if st == nil {
		return false
	}
	st.mu.Lock()
	defer st.mu.Unlock()
	return st.initialized
}

// String returns the current value (non-secret), or "" before Init.
func (p ParameterValue) String() string { return p.Get() }
